Add a page on physical light units (#6390)
godotengine/godot-docs: tutorials/3d/physical_light_and_camera_units.rst

.. _doc_physical_light_and_camera_units:

Physical light and camera units
===============================

Why use physical light and camera units?
----------------------------------------

> Since video game graphics generally don't aim for realism, Godot uses arbitrary
> units by default to favor usability. This applies to light color and energy,
> camera field of view, exposure, and more.

Reviewer: ```suggestion
Godot uses arbitrary units for many physical properties that apply to light
like color, energy, camera field of view, and exposure. By default these
properties use arbitrary units because using accurate physical units comes
with a few tradeoffs that are not worth making for many games. As Godot
favors ease-of-use by default, physical light units are disabled by default.
```

Advantages
^^^^^^^^^^

If you aim for photorealism in your project, using real world units as a basis
can help make things easier to adjust. References for real world materials,

> lights and scene brightness are wildly available online. This can also be useful

Reviewer: We could include a link to https://physicallybased.info/ here

when porting a scene from other 3D software that uses physical light units (such
as Blender).

Disadvantages
^^^^^^^^^^^^^

Using physical light and camera units will not automatically make your project
look *better*. Sometimes, moving away from realism can actually make a scene
look better to the human eye. Also, using physical units also requires a greater
amount of rigor compared to non-physical units. Most benefits of physical units
can only be obtained if the units are correctly set to match real world

> reference.
> 

Reviewer: ```suggestion
reference. The biggest disadvantage of using physical light units is you will have to pay
close attention to the dynamic range in use at a given time as you can run into floating
point precision errors when mixing very high light intensities with very low light intensities. 
In practice, this means that you will have to manually manage your exposure settings to
ensure that you are not over-exposing or under-exposing your scene by too much. 
Auto-exposure can help you balance the light in a scene to bring it into a normal range, 
but it cannot recover lost precision from a dynamic range that is too high.

```

.. note::

    Physical light units are only available in 3D rendering, not 2D.

Setting up physical light units
-------------------------------

Physical light units can be enabled separately from physical camera units.

To enable physical light units correctly, there are 4 steps required:

1. Enable the project setting.
2. Configure the camera.
3. Configure the environment.
4. Configure Light3D nodes.

Since physical light and camera units only require a handful of calculations to
handle unit conversion, enabling them doesn't have any noticeable performance

> impact.

Author: This is technically not true, as depth of field is enforced when CameraAttributesPhysical is used. Is there a way to disable depth of field entirely when using CameraAttributesPhysical?

Reviewer: There isn't a way to force disable DoF in CameraAttributesPhysical right now. It is enabled/disabled automatically based on the parameters used

Author: I suppose this could be implemented as part of https://github.com/godotengine/godot-proposals/issues/5229.

Enable the project setting
^^^^^^^^^^^^^^^^^^^^^^^^^^

Open the Project Settings, enable the **Advanced** toggle then enable
**Rendering > Lights And Shadows > Use Physical Light Units**. Restart the editor.

Configure the camera
^^^^^^^^^^^^^^^^^^^^

.. warning::

>     When physical light units are enabled, you **must** have a
>     :ref:`class_CameraAttributes` resource assigned to the
>     WorldEnvironment node. Otherwise, the 3D editor viewport will
>     appear extremely bright if you have a visible DirectionalLight3D node.

Reviewer: ```suggestion
    When physical light units are enabled and if you have a
    WorldEnvironment node in your scene (i.e. the editor
    Environment is disabled), you **must** have a
    :ref:`class_CameraAttributes` resource assigned to the
    WorldEnvironment node. Otherwise, the 3D editor viewport will
    appear extremely bright if you have a visible DirectionalLight3D node.
```

The editor environment is tuned to work with the editor light, so if they have both present it will look fine.

outdoor photography depending on weather conditions. Higher values are used for
indoor or night-time photography.

.. note::

    Unlike real life cameras, the adverse effects of increasing ISO sensitivity
    or decreasing shutter speed (such as visible grain or light trails) are not
    simulated in Godot.

See :ref:`doc_physical_light_and_camera_units_setting_up_physical_camera_units`
for a description of CameraAttributesPhysical properties that are also available when
**not** using physical light units.

Configure the environment
^^^^^^^^^^^^^^^^^^^^^^^^^

.. warning::

    The default configuration is designed for daytime outdoor scenes. Night-time
    and indoor scenes will need adjustments to the DirectionalLight3D and
    WorldEnvironment background intensity to look correct. Otherwise, positional
    lights will be barely visible at their default intensity.

If you haven't added a :ref:`class_WorldEnvironment` and :ref:`class_Camera3D`
node to the current scene yet, do so now by clicking the 3 vertical dots at the
top of the 3D editor viewport. Click **Add Sun to Scene**, open the dialog again
then click **Add Environment to Scene**.

> After enabling physical light unit, a new property becomes available to edit in

Reviewer: ```suggestion
After enabling physical light units, a new property becomes available to edit in
```
